File: native/collectors/safety/sources/jma_warnings.c
/* collectors/safety/sources/jma_warnings.c
 * Port of server/src/collectors/jmaWarnings.js — JMA bosai warning JSON,
 * one file per prefecture office code. Emit one geo point per area that
 * currently has >=1 active warning (status != "解除", code != "00").
 * Real JMA data only; honest empty when nothing fetched (rule 8).
 * Property key order mirrors the JS mapFn for featureUid hash parity. */
#include "../../../source.h"
#include "../../../lib/feedlib.h"
#include "../../../lib/geojson.h"
#include "../../../third_party/cJSON.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define WARN_URL "https://www.jma.go.jp/bosai/warning/data/warning/"

struct area { const char *code, *name; double lat, lon; };
static const struct area AREAS[] = {
  {"016000","北海道（石狩・空知・後志）",43.06,141.35},
  {"020000","青森県",40.82,140.74}, {"030000","岩手県",39.70,141.15},
  {"040000","宮城県",38.27,140.87}, {"050000","秋田県",39.72,140.10},
  {"060000","山形県",38.24,140.36}, {"070000","福島県",37.75,140.47},
  {"080000","茨城県",36.37,140.47}, {"090000","栃木県",36.57,139.88},
  {"100000","群馬県",36.39,139.06}, {"110000","埼玉県",35.86,139.65},
  {"120000","千葉県",35.61,140.12}, {"130000","東京都",35.69,139.69},
  {"140000","神奈川県",35.45,139.64}, {"150000","新潟県",37.90,139.02},
  {"160000","富山県",36.70,137.21}, {"170000","石川県",36.59,136.63},
  {"180000","福井県",36.07,136.22}, {"190000","山梨県",35.66,138.57},
  {"200000","長野県",36.65,138.18}, {"210000","岐阜県",35.39,136.72},
  {"220000","静岡県",34.98,138.38}, {"230000","愛知県",35.18,136.91},
  {"240000","三重県",34.73,136.51}, {"250000","滋賀県",35.00,135.87},
  {"260000","京都府",35.02,135.76}, {"270000","大阪府",34.69,135.52},
  {"280000","兵庫県",34.69,135.18}, {"290000","奈良県",34.69,135.83},
  {"300000","和歌山県",34.23,135.17}, {"310000","鳥取県",35.50,134.24},
  {"320000","島根県",35.47,133.05}, {"330000","岡山県",34.66,133.93},
  {"340000","広島県",34.40,132.46}, {"350000","山口県",34.19,131.47},
  {"360000","徳島県",34.07,134.56}, {"370000","香川県",34.34,134.04},
  {"380000","愛媛県",33.84,132.77}, {"390000","高知県",33.56,133.53},
  {"400000","福岡県",33.59,130.40}, {"410000","佐賀県",33.25,130.30},
  {"420000","長崎県",32.74,129.87}, {"430000","熊本県",32.80,130.73},
  {"440000","大分県",33.24,131.61}, {"450000","宮崎県",31.91,131.42},
  {"460100","鹿児島県",31.56,130.56}, {"471000","沖縄県（本島中南部）",26.21,127.68},
};
#define NAREA ((int)(sizeof AREAS / sizeof AREAS[0]))

/* Collect distinct active warning names from a bosai warning doc. */
static cJSON *extract_active(cJSON *doc) {
  cJSON *out = cJSON_CreateArray();
  cJSON *areaTypes = cJSON_GetObjectItem(doc, "areaTypes");
  if (!areaTypes || !cJSON_IsArray(areaTypes)) return out;
  cJSON *at;
  cJSON_ArrayForEach(at, areaTypes) {
    cJSON *areas = cJSON_GetObjectItem(at, "areas");
    if (!areas || !cJSON_IsArray(areas)) continue;
    cJSON *a;
    cJSON_ArrayForEach(a, areas) {
      cJSON *warns = cJSON_GetObjectItem(a, "warnings");
      if (!warns || !cJSON_IsArray(warns)) continue;
      cJSON *w;
      cJSON_ArrayForEach(w, warns) {
        cJSON *st = cJSON_GetObjectItem(w, "status");
        cJSON *cd = cJSON_GetObjectItem(w, "code");
        if (!st || !cJSON_IsString(st)) continue;
        if (strcmp(st->valuestring, "解除") == 0) continue;
        if (!cd || !cJSON_IsString(cd)) continue;
        if (strcmp(cd->valuestring, "00") == 0) continue;
        cJSON *nm = cJSON_GetObjectItem(w, "name");
        const char *val = (nm && cJSON_IsString(nm)) ? nm->valuestring
                                                     : cd->valuestring;
        /* unique */
        int dup = 0; cJSON *e;
        cJSON_ArrayForEach(e, out)
          if (cJSON_IsString(e) && strcmp(e->valuestring, val) == 0) { dup = 1; break; }
        if (!dup) cJSON_AddItemToArray(out, cJSON_CreateString(val));
      }
    }
  }
  return out;
}

static int run(const source_ctx *ctx, intel_sink *sink) {
  cJSON *features = cJSON_CreateArray();
  int any_fetched = 0;

  for (int i = 0; i < NAREA; i++) {
    char url[160];
    snprintf(url, sizeof url, "%s%s.json", WARN_URL, AREAS[i].code);
    cJSON *doc = feed_get_json(ctx->http, url, 8000);
    if (!doc) continue;
    any_fetched = 1;

    cJSON *warns = extract_active(doc);
    if (cJSON_GetArraySize(warns) == 0) { cJSON_Delete(warns); cJSON_Delete(doc); continue; }

    cJSON *rd = cJSON_GetObjectItem(doc, "reportDatetime");

    cJSON *f = cJSON_CreateObject();
    cJSON_AddStringToObject(f, "type", "Feature");
    cJSON *g = cJSON_CreateObject();
    cJSON_AddStringToObject(g, "type", "Point");
    cJSON *c = cJSON_CreateArray();
    cJSON_AddItemToArray(c, cJSON_CreateNumber(AREAS[i].lon));
    cJSON_AddItemToArray(c, cJSON_CreateNumber(AREAS[i].lat));
    cJSON_AddItemToObject(g, "coordinates", c);
    cJSON_AddItemToObject(f, "geometry", g);

    cJSON *p = cJSON_CreateObject();              /* EXACT JS key order */
    cJSON_AddStringToObject(p, "area_code", AREAS[i].code);
    cJSON_AddStringToObject(p, "area_name", AREAS[i].name);
    cJSON_AddItemToObject(p, "warnings", warns);
    cJSON_AddNumberToObject(p, "warning_count", cJSON_GetArraySize(warns));
    if (rd && cJSON_IsString(rd)) cJSON_AddStringToObject(p, "report_at", rd->valuestring);
    else                          cJSON_AddNullToObject(p, "report_at");
    cJSON_AddStringToObject(p, "source", "jma_warning");
    cJSON_AddItemToObject(f, "properties", p);
    cJSON_AddItemToArray(features, f);
    cJSON_Delete(doc);
  }

  if (!any_fetched) { cJSON_Delete(features); fprintf(stderr, "[jma-warnings] unavailable\n"); return -1; }

  int n = geojson_emit_features(sink, ctx->source_id, features);
  cJSON_Delete(features);
  fprintf(stderr, "[jma-warnings] emitted %d\n", n);
  return n >= 0 ? 0 : -1;
}

static const source_def jma_warnings_def = {
  .id = "jma-warnings", .collector = "safety",
  .name = "JMA Weather Warnings", .name_ja = "気象庁 気象警報",
  .update_interval_sec = 300, .run = run };
REGISTER_SOURCE(jma_warnings_def)

File: server/src/collectors/jmaWarnings.js
/**
 * JMA weather warnings / advisories — jma-warnings.
 *
 * Real: JMA bosai warning JSON, one file per prefecture area code:
 *   https://www.jma.go.jp/bosai/warning/data/warning/{areaCode}.json
 * Analogous to jmaForecastArea.js. We iterate the prefecture office codes,
 * emit one point per area that currently has at least one active warning
 * (status != type "解除"/empty). No fabrication: on failure → honest empty.
 */

// Prefecture forecast-office codes with a representative office coordinate.
const AREAS = [
  { code: '016000', name: '北海道（石狩・空知・後志）', lat: 43.06, lon: 141.35 },
  { code: '020000', name: '青森県', lat: 40.82, lon: 140.74 },
  { code: '030000', name: '岩手県', lat: 39.70, lon: 141.15 },
  { code: '040000', name: '宮城県', lat: 38.27, lon: 140.87 },
  { code: '050000', name: '秋田県', lat: 39.72, lon: 140.10 },
  { code: '060000', name: '山形県', lat: 38.24, lon: 140.36 },
  { code: '070000', name: '福島県', lat: 37.75, lon: 140.47 },
  { code: '080000', name: '茨城県', lat: 36.37, lon: 140.47 },
  { code: '090000', name: '栃木県', lat: 36.57, lon: 139.88 },
  { code: '100000', name: '群馬県', lat: 36.39, lon: 139.06 },
  { code: '110000', name: '埼玉県', lat: 35.86, lon: 139.65 },
  { code: '120000', name: '千葉県', lat: 35.61, lon: 140.12 },
  { code: '130000', name: '東京都', lat: 35.69, lon: 139.69 },
  { code: '140000', name: '神奈川県', lat: 35.45, lon: 139.64 },
  { code: '150000', name: '新潟県', lat: 37.90, lon: 139.02 },
  { code: '160000', name: '富山県', lat: 36.70, lon: 137.21 },
  { code: '170000', name: '石川県', lat: 36.59, lon: 136.63 },
  { code: '180000', name: '福井県', lat: 36.07, lon: 136.22 },
  { code: '190000', name: '山梨県', lat: 35.66, lon: 138.57 },
  { code: '200000', name: '長野県', lat: 36.65, lon: 138.18 },
  { code: '210000', name: '岐阜県', lat: 35.39, lon: 136.72 },
  { code: '220000', name: '静岡県', lat: 34.98, lon: 138.38 },
  { code: '230000', name: '愛知県', lat: 35.18, lon: 136.91 },
  { code: '240000', name: '三重県', lat: 34.73, lon: 136.51 },
  { code: '250000', name: '滋賀県', lat: 35.00, lon: 135.87 },
  { code: '260000', name: '京都府', lat: 35.02, lon: 135.76 },
  { code: '270000', name: '大阪府', lat: 34.69, lon: 135.52 },
  { code: '280000', name: '兵庫県', lat: 34.69, lon: 135.18 },
  { code: '290000', name: '奈良県', lat: 34.69, lon: 135.83 },
  { code: '300000', name: '和歌山県', lat: 34.23, lon: 135.17 },
  { code: '310000', name: '鳥取県', lat: 35.50, lon: 134.24 },
  { code: '320000', name: '島根県', lat: 35.47, lon: 133.05 },
  { code: '330000', name: '岡山県', lat: 34.66, lon: 133.93 },
  { code: '340000', name: '広島県', lat: 34.40, lon: 132.46 },
  { code: '350000', name: '山口県', lat: 34.19, lon: 131.47 },
  { code: '360000', name: '徳島県', lat: 34.07, lon: 134.56 },
  { code: '370000', name: '香川県', lat: 34.34, lon: 134.04 },
  { code: '380000', name: '愛媛県', lat: 33.84, lon: 132.77 },
  { code: '390000', name: '高知県', lat: 33.56, lon: 133.53 },
  { code: '400000', name: '福岡県', lat: 33.59, lon: 130.40 },
  { code: '410000', name: '佐賀県', lat: 33.25, lon: 130.30 },
  { code: '420000', name: '長崎県', lat: 32.74, lon: 129.87 },
  { code: '430000', name: '熊本県', lat: 32.80, lon: 130.73 },
  { code: '440000', name: '大分県', lat: 33.24, lon: 131.61 },
  { code: '450000', name: '宮崎県', lat: 31.91, lon: 131.42 },
  { code: '460100', name: '鹿児島県', lat: 31.56, lon: 130.56 },
  { code: '471000', name: '沖縄県（本島中南部）', lat: 26.21, lon: 127.68 },
];
const TIMEOUT_MS = 8000;

async function fetchOne(area) {
  try {
    const ctrl = new AbortController();
    const timer = setTimeout(() => ctrl.abort(), TIMEOUT_MS);
    const res = await fetch(
      `https://www.jma.go.jp/bosai/warning/data/warning/${area.code}.json`,
      { signal: ctrl.signal },
    );
    clearTimeout(timer);
    if (!res.ok) return null;
    const data = await res.json();
    return { area, data };
  } catch {
    return null;
  }
}

// Collect active warning type names from a bosai warning document.
function extractActiveWarnings(doc) {
  const out = [];
  const areaTypes = doc?.areaTypes;
  if (!Array.isArray(areaTypes)) return out;
  for (const at of areaTypes) {
    for (const a of at.areas ?? []) {
      for (const w of a.warnings ?? []) {
        // status of "発表" / "継続" etc. means active; "解除" means cleared.
        if (w.status && w.status !== '解除' && w.code && w.code !== '00') {
          out.push(w.name || w.code);
        }
      }
    }
  }
  return Array.from(new Set(out));
}

function emptyResult(source) {
  return {
    type: 'FeatureCollection',
    features: [],
    _meta: {
      source,
      fetchedAt: new Date().toISOString(),
      recordCount: 0,
      description: 'JMA active weather warnings / advisories per prefecture',
    },
  };
}

export default async function collectJmaWarnings() {
  let results;
  try {
    results = await Promise.all(AREAS.map(fetchOne));
  } catch {
    return emptyResult('jma_warnings_unavailable');
  }

  const features = [];
  let anyFetched = false;
  for (const r of results) {
    if (!r || !r.data) continue;
    anyFetched = true;
    const warnings = extractActiveWarnings(r.data);
    if (warnings.length === 0) continue;
    features.push({
      type: 'Feature',
      geometry: { type: 'Point', coordinates: [r.area.lon, r.area.lat] },
      properties: {
        area_code: r.area.code,
        area_name: r.area.name,
        warnings,
        warning_count: warnings.length,
        report_at: r.data?.reportDatetime ?? null,
        source: 'jma_warning',
      },
    });
  }

  if (!anyFetched) return emptyResult('jma_warnings_unavailable');

  return {
    type: 'FeatureCollection',
    features,
    _meta: {
      source: 'jma_warning',
      fetchedAt: new Date().toISOString(),
      recordCount: features.length,
      description: 'JMA active weather warnings / advisories per prefecture',
    },
  };
}
